File: 21. Merge Two Sorted Lists.rb
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end

list1 = ListNode.new(5, ListNode.new(9, ListNode.new(12, nil)))
list2 = ListNode.new(1, ListNode.new(5, ListNode.new(7, ListNode.new(10, nil))))


def merge_two_lists(list1, list2)
  final_list = list1
  current = list2

  if !list1 && list2
    return list2
  
  elsif !list2 && list1
    return list1
  
  elsif !list1 && !list2
    return nil
  end

  while current
    final_list = add_node(final_list, current.val)
    current = current.next
  end

  return final_list
end

# function to add new node at the right place
def add_node(list, value)
  current = list
  prev = nil

  while current
    
    # case-1:
    # if the right place of the node is at first of the list
    if prev === nil
      if value < current.val
        new_node = ListNode.new(value, current)
        return new_node
      end
    end

    # case-2:
    # if the right place of the node is between nodes
    if value < current.val
      new_node = ListNode.new(value, current)
      prev.next = new_node
      return list
    end

    # case-3:
    # if the right place of the node is at last of the list
    if !current.next
      new_node = ListNode.new(value, nil)
      current.next = new_node
      return list
    end

    # go to the next node
    prev = current
    current = current.next

  end
end

puts merge_two_lists(list1, list2).inspect

File: 21. Merge Two Sorted Lists.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function mergeTwoLists(
  list1: ListNode | null,
  list2: ListNode | null
): ListNode | null {
  if (!list1) return list2;
  else if (!list2) return list1;
  else if (!list1 && !list2) return null;

  let list1Head: ListNode | null = list1;
  let list2Head: ListNode | null = list2;
  let newList = new ListNode();
  let newListHead = newList;

  while (list1Head || list2Head) {
    if (list1Head && list2Head) {
      if (list1Head.val <= list2Head.val) {
        newListHead.val = list1Head.val;
        list1Head = list1Head.next;
      } else if (list1Head.val > list2Head.val) {
        newListHead.val = list2Head.val;
        list2Head = list2Head.next;
      }
    } else if (list1Head) {
      newListHead.val = list1Head.val;
      list1Head = list1Head.next;
    } else if (list2Head) {
      newListHead.val = list2Head.val;
      list2Head = list2Head.next;
    }

    if (list1Head || list2Head) {
      newListHead.next = new ListNode();
      newListHead = newListHead.next;
    }
  }

  return newList;
}
